ssz/src/main/java/tech/pegasys/teku/ssz/backing/tree/SszSuperNode.java

```diff
@@ -0,0 +1,172 @@
+  }
+
+  private Bytes32 hashTreeRoot(int curDepth, int offset) {
+    if (curDepth == depth) {
+      if (offset < ssz.size()) {
+        return elementTemplate.calculateHashTreeRoot(ssz, offset);
+      } else {
+        return DEFAULT_NODE.hashTreeRoot();
+      }
+    } else {
+      return Hash.sha2_256(
+          Bytes.wrap(
+              hashTreeRoot(curDepth + 1, offset),
+              hashTreeRoot(
+                  curDepth + 1,
+                  offset + elementTemplate.getSszLength() * (1 << ((depth - curDepth) - 1)))));
+    }
+  }
+
+  @NotNull
+  @Override
+  public TreeNode get(long generalizedIndex) {
+    if (gIdxIsSelf(generalizedIndex)) {
+      return this;
+    }
+    int childIndex = gIdxGetChildIndex(generalizedIndex, depth);
+    int childOffset = childIndex * elementTemplate.getSszLength();
+    checkArgument(childOffset < ssz.size(), "Invalid index");
+    long relativeGIndex = gIdxGetRelativeGIndex(generalizedIndex, depth);
+    Location nodeLoc = elementTemplate.getNodeSszLocation(relativeGIndex);
+    if (nodeLoc.isLeaf()) {
+      return LeafNode.create(ssz.slice(childOffset + nodeLoc.getOffset(), nodeLoc.getLength()));
+    } else if (gIdxIsSelf(relativeGIndex)) {
+      return new SszSuperNode(
+          0, elementTemplate, ssz.slice(childOffset, elementTemplate.getSszLength()));
+    } else {
+      SszNodeTemplate subTemplate = elementTemplate.getSubTemplate(relativeGIndex);
+      return new SszSuperNode(
+          0, subTemplate, ssz.slice(childOffset + nodeLoc.getOffset(), nodeLoc.getLength()));
+    }
+  }
+
+  @Override
+  public boolean iterate(
+      long thisGeneralizedIndex, long startGeneralizedIndex, TreeVisitor visitor) {
+    if (gIdxCompare(thisGeneralizedIndex, startGeneralizedIndex) == NodeRelation.Left) {
+      return true;
+    } else {
+      return visitor.visit(this, thisGeneralizedIndex);
+    }
+  }
+
+  @Override
+  public TreeNode updated(TreeUpdates newNodes) {
+    if (newNodes.isEmpty()) {
+      return this;
+    }
+    long leftmostUpdateIndex = newNodes.getRelativeGIndex(newNodes.size() - 1);
+    int leftmostChildIndex = gIdxGetChildIndex(leftmostUpdateIndex, depth);
+    int newSszSize = (leftmostChildIndex + 1) * elementTemplate.getSszLength();
+    Bytes updatedSizeSsz =
+        newSszSize <= ssz.size()
+            ? ssz
+            : Bytes.wrap(ssz, Bytes.wrap(new byte[newSszSize - ssz.size()]));
+    MutableBytes mutableCopy = updatedSizeSsz.mutableCopy();
+    for (int i = 0; i < newNodes.size(); i++) {
+      long updateGIndex = newNodes.getRelativeGIndex(i);
+      int childIndex = gIdxGetChildIndex(updateGIndex, depth);
+      long childGIndex = gIdxGetRelativeGIndex(updateGIndex, depth);
+      int childOffset = childIndex * elementTemplate.getSszLength();
+      MutableBytes childMutableSlice =
+          mutableCopy.mutableSlice(childOffset, elementTemplate.getSszLength());
+      elementTemplate.update(childGIndex, newNodes.getNode(i), childMutableSlice);
```

Comment:
Would it make more sense to do the ssz update on the SszSuperNode object rather than the template? To me the template seems to almost exclusively deal with the structure/location of the trees

Reply:
Template (`elementTemplate`) here deals with the structure of the list/vector element while the `SszSuperNode` deals with a 'set' of elements. I've tried to move as much as possible of functionality specific to just element type to this template. I.e. a `Node` holds actual data and its template knows how to interpret it.

I was even going to make another template class which would represent a set of elements (like `SszSuperNode` does now) and leave the `SszSuperNode` with even more weak functionality. But just stuck with this

This is how an 'ideal' node implementation would look like for me: https://gist.github.com/Nashatyrev/8f416e4d3be10347a7c5f6c6520b1f53